### ISSUE TITLE
Microsoft.CodeAnalysis.NetAnalyzers not doing anything at all for me

### ISSUE THREAD

**Issue by endasil** [dotnet/roslyn-analyzers#4776]
So I switched from using the fxcop nuget to the new Microsoft.CodeAnalysis.NetAnalyzers NuGet and it does nothing at all. I added the following to my csproj file expecting it to apply all rules but nothing happens when i build. I uploaded an example project https://easyupload.io/iktg0k 

I added the following to my csproj file
<PropertyGroup>
    <EnableNETAnalyzers>true</EnableNETAnalyzers>
    <AnalysisLevel>latest</AnalysisLevel>
    <AnalysisMode>AllEnabledByDefault</AnalysisMode>
  </PropertyGroup>

my application looks like this and should generate https://docs.microsoft.com/en-us/dotnet/fundamentals/code-analysis/quality-rules/ca1819 but does not seem to generate any warnings or messages at all

    namespace ConsoleApp1
    {
        public class Book
        {
            private string[] _Pages;
    
            public Book(string[] pages)
            {
                _Pages = pages;
            }
    
            public string[] Pages
            {
                get { return _Pages; }
            }
        }
    
        class Program
        {
            static void Main(string[] args)
            {
            }
        }
    }

**Comment by mavasani** [comment]
Seems to work fine for a vanilla net5 classlib project for me:

![image](https://user-images.githubusercontent.com/10605811/106818258-e286a080-662c-11eb-9915-c925f07117c4.png)

**Comment by endasil** [comment]
Right, how about with a net472 project?

**Comment by Youssef1313** [comment]
@endasil Usually it should work (I've tried that before, but using the SDK-style project files).

I can confirm that it doesn't work for the project you attached (it could be something related to the old csproj format?):

![image](https://user-images.githubusercontent.com/31348972/106886663-cfd3a080-66ec-11eb-8621-badd1bd686f9.png)

**Comment by Youssef1313** [comment]
Hmmm, I could reproduce even with SDK-style project.

![image](https://user-images.githubusercontent.com/31348972/106888835-cac42080-66ef-11eb-8150-c7ae5d20f1d9.png)


[netanalyserfail.zip](https://github.com/dotnet/roslyn-analyzers/files/5925502/netanalyserfail.zip)

**Comment by Youssef1313** [comment]
@enasil One problem with your repro (the zip file - not the code in issue) is that `Pages` property is **not** in the public api surface (the containing class `Program` is internal):

![image](https://user-images.githubusercontent.com/31348972/106889958-6609c580-66f1-11eb-836f-870987b6dea2.png)

After fixing that, I still don't get the diagnostic. Only after adding an editorconfig with the following content I get the diagnostic:

```ini
[*.cs]
dotnet_diagnostic.CA1819.severity = warning
```

I'll give that a look.

**Comment by Youssef1313** [comment]
@mavasani editorconfigs from package are not included in build when the csproj isn't SDK-style. Is this by-design?

**Comment by mavasani** [comment]
Yes. For legacy csproj style projects, the only option is to reference the analyzer NuGet package. We should reflect this in docs.

**Comment by Youssef1313** [comment]
@mavasani The package is already installed:

![image](https://user-images.githubusercontent.com/31348972/106899839-7f653e80-66fe-11eb-88b3-d33ae6cd6ff6.png)

**Comment by mavasani** [comment]
There could be a different issue then, would need investigation.

**Comment by Youssef1313** [comment]
Looking at the binlog, I can't find the editorconfigs from package. Once I switch to SDK-style, I can see them.

Here is the binary log:


[msbuild.zip](https://github.com/dotnet/roslyn-analyzers/files/5926071/msbuild.zip)

**Comment by mavasani** [comment]
@Youssef1313 The logic that maps the `AnalysisMode` and `AnalysisLevel` property to correct global config file relies on some MSBuild logic that is in the SDK, so these properties will not work with legacy csproj style projects. @gewarren We should make this explicit in our documentation.

**Comment by stewartadam** [comment]
I'm facing the same issue. A project targeting `netcoreapp3.1` with `Microsoft.CodeAnalysis.NetAnalyzers` added does not produce the same warnings when compiled with .NET Core 3.1 SDK vs .NET 5 SDK:

```
<Project Sdk="Microsoft.NET.Sdk.Web">

  <PropertyGroup>
    <TargetFramework>netcoreapp3.1</TargetFramework>
    <EnableNETAnalyzers>false</EnableNETAnalyzers>
    <EnforceCodeStyleInBuild>true</EnforceCodeStyleInBuild>
    <AnalysisLevel>latest</AnalysisLevel>
    <AnalysisMode>Recommended</AnalysisMode>
  </PropertyGroup>
  ...
  <ItemGroup>
    <PackageReference Include="Microsoft.CodeAnalysis.NetAnalyzers" Version="5.0.3">
            <PrivateAssets>all</PrivateAssets>
            <IncludeAssets>runtime; build; native; contentfiles; analyzers; buildtransitive</IncludeAssets>
        </PackageReference>
  </ItemGroup>
```

The value of `Recommended` is not documented but works since the NuGet package ships with editorconfigs for it. It doesn't matter though, setting it to the documented values of `AllEnabledByDefault` has no effect either when compiling with .NET Core 3.1 SDK.

**Comment by xiaomi7732** [comment]
It is NOT working for me either when using 3.1 SDK (5.0 works). Build succeeded without warning or error.
I have a repro repository here, can someone help take a look?

<https://github.com/xiaomi7732/NetAnalyzerCore3_1Repro>

1. I have global.json to pick 3.1 SDK (again, set it 5.0 will work, I can see warnings):

```json
{
  "sdk": {
    "version": "3.1.408"
  }
}
```

2. This is a .NET Core console app, project properties like this:

```xml
<Project Sdk="Microsoft.NET.Sdk">

  <PropertyGroup>
    <OutputType>Exe</OutputType>
    <TargetFramework>netcoreapp3.1</TargetFramework>
    <!-- <TreatWarningsAsErrors>true</TreatWarningsAsErrors> -->
    <EnableNETAnalyzers>true</EnableNETAnalyzers>
    <AnalysisMode>AllEnabledByDefault</AnalysisMode>
    <AnalysisLevel>latest</AnalysisLevel>
  </PropertyGroup>

  <ItemGroup>
    <PackageReference Include="Microsoft.CodeAnalysis.NetAnalyzers" Version="5.0.3">
      <IncludeAssets>runtime; build; native; contentfiles; analyzers; buildtransitive</IncludeAssets>
      <PrivateAssets>all</PrivateAssets>
    </PackageReference>
  </ItemGroup>

</Project>
```

**Comment by Youssef1313** [comment]
@xiaomi7732 Your issue may be different than the original one. Can you open a new issue for investigation?

**Comment by xiaomi7732** [comment]
@Youssef1313 New issue filed: #5069

**Comment by segevfiner** [comment]
So basically some logic is missing from the msbuild targets for this to work on non-SDK projects. e.g. A Visual Studio/VSIX extension project.

 It would be nice if this was fixed as that means you are stuck with legacy unmaintained analyzers for such a project type...

**Comment by cremor** [comment]
The information provided by @segevfiner in the previous comment seems to be the only working configuration, thanks!

Why is there no proper documentation for this?
Or better: Why is the documentation even stating that `AnalysisMode` doesn't work for the Microsoft.CodeAnalysis.NetAnalyzers NuGet package, even though it clearly does (and is required)? See "note" block on https://learn.microsoft.com/en-us/dotnet/core/project-sdk/msbuild-props#analysismode

**Comment by gewarren** [comment]
Thanks @segevfiner. Docs updated in https://github.com/dotnet/docs/pull/48804.

**Comment by cremor** [comment]
Does anyone know how to get analysis level 10 working in legacy csproj files?

When I update the Microsoft.CodeAnalysis.NetAnalyzers package from version 9.0.0 to version 10.0.203 then the `AnalysisMode` configuration doesn't work any more. It looks like it always uses `Default`.

This is with the legacy csproj elements configured as shown by @segevfiner above, and as is now officially [documented](https://learn.microsoft.com/en-us/dotnet/fundamentals/code-analysis/overview#enable-code-analysis-in-legacy-projects).

Maybe the problem is that the 10.x NuGet packages don't include configuration files any more?
You can see this if you compare https://nuget.info/packages/Microsoft.CodeAnalysis.NetAnalyzers/9.0.0 and https://nuget.info/packages/Microsoft.CodeAnalysis.NetAnalyzers/10.0.203. The config folder `buildTransitive\config` (and other folders) are missing in the 10.x NuGet package versions.
See also dotnet/sdk#54046
